designs/aws/custom-user-data-and-amis.md

```diff
@@ -0,0 +1,285 @@
+    * This helps w/ the bin-packing logic since it will probably be impossible to parse the user-data as a bash script.
+* Users can then address CASE2 for most cases. But this still leaves the door open for everyone to try and add more kubelet and container configuration within the provisioner which will bloat our API surface and make the provisioner more challenging to maintain.
+
+*[Recommended] Option 3 - MIME parts are run in the order - `Karpenter Managed` → `UserManaged` → `Karpenter Managed`*
+
+```
+MIME-Version: 1.0
+Content-Type: multipart/mixed; boundary="==MYBOUNDARY=="
+
+--==MYBOUNDARY==
+Content-Type: text/x-shellscript; charset="us-ascii"
+
+#!/bin/bash
+set -ex
+B64_CLUSTER_CA=base64data
+API_SERVER_URL=apiServerEndpoint
+K8S_CLUSTER_DNS_IP=k8sClusterDnsIp
+/etc/eks/bootstrap.sh clusterName
+  --kubelet-extra-args '--node-labels=karpenter.sh/provisioner-name=default,karpenter.sh/capacity-type=on-demand' \
+  --b64-cluster-ca $B64_CLUSTER_CA \
+  --apiserver-endpoint $API_SERVER_URL \
+  --dns-cluster-ip $K8S_CLUSTER_DNS_IP \
+  --kubelet-disabled
+
+--==MYBOUNDARY==
+Content-Type: text/x-shellscript; charset="us-ascii"
+#!/bin/bash
+
+# Running custom user data script that was in the ProvisionerSpec
+# At this point, the kubelet hasn't started so as a user I can make mutations
+# to any of the kubeletConfig / containerRuntime as needed.
+
+# To edit kubelet extra-args - /etc/systemd/system/kubelet.service.d/30-kubelet-extra-args.conf
+# To edit other kubelet args - /etc/systemd/system/kubelet.service.d/
+# To edit kubelet-config - /etc/kubernetes/kubelet/kubelet-config.json
+
+
+--==MYBOUNDARY==
+Content-Type: text/x-shellscript; charset="us-ascii"
+
+#!/bin/bash
+systemctl enable kubelet
+systemctl restart kubelet
+
+--==MYBOUNDARY==--
+```
+
+* In this mode, Karpenter passes a new parameter to the bootstrap script called `kubelet-disabled`. The bootstrap script will function normally, except the kubelet systemd service will not be started. This final hook gives Karpenter users time to make any modifications they need to make.
+* Making modifications via bash in the UserData is *not easy*. Using jq _*is painful*_ because it doesn’t support in-place updates / json merges etc.
+    * Updating the `.conf` files that has the kubelet args is difficult since you'd need to sed / awk your way through.
+* Your options are now limitless - you have full control over the worker node's contents and can potentially make any changes you'd like. You can also see what Karpenter was planning to start the kubelet as.
+* We’re establishing a runtime-contract that the the kubelet configurations will always be in-place at specific file locations. We can’t change this runtime contract over time for backwards breaking compatibility.
```

Comment:
I think you probably meant "backwards compatibility" not "backwards breaking compatibility"